src/file.rs

```diff
@@ -1,117 +1,556 @@
-use std::ffi;
-use std::path;
-use std::collections::HashMap;
+use super::error;
+use super::group::Group;
+use super::HashMap;
+use super::LOCK;
 use netcdf_sys::*;
-use group::{init_group, Group};
-use NC_ERRORS;
+use std::ffi::CString;
+use std::path;
 
+#[derive(Debug)]
 pub struct File {
-    pub id: i32,
-    pub name: String,
-    pub root: Group,
+    pub(crate) ncid: nc_type,
+    pub(crate) name: String,
+    pub(crate) root: Group,
+}
+
+impl File {
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    /// Main entrypoint for interacting with the netcdf file. Also accesible
+    /// through the `Deref` trait on `File`
+    pub fn root(&self) -> &Group {
+        &self.root
+    }
+
+    /// Mutable access to the root group
+    pub fn root_mut(&mut self) -> &mut Group {
+        &mut self.root
+    }
+}
+
+impl std::ops::Deref for File {
+    type Target = Group;
+    fn deref(&self) -> &Self::Target {
+        &self.root
+    }
+}
+
+impl std::ops::DerefMut for File {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.root
+    }
+}
+
+impl File {
+    /// Open a netCDF file in read only mode.
+    pub fn open<P>(file: P) -> error::Result<File>
+    where
+        P: AsRef<path::Path>,
+    {
```

Comment:
I'm just reading through the PR, it's not really my competence to review it. I can give some generic hints and it's just intended as commentary, you can judge what to implement.

It looks like open, append, create are all good candidates for being less generic. The pattern would be that `pub fn open<P>(file: P) -> error::Result<File>` remains as it is, but inside it just calls `open_impl(file.as_ref())`. This is effective when `open_impl` is non-generic non-inline and has a substantial body; it speeds up compilation of crates that use this crate.

Reply:
I agree! The natural entrypoints will be netcdf::open/append/create, which just acts as shims for File::{open,append,create} anyways